backend/notification/rate_limiter.js

```diff
@@ -1,34 +1,119 @@
+const prisma = require("../prisma/prisma_client");
+
 class RateLimiter {
-  constructor(time_window, maximum_requests) {
-    this.time_window = time_window * 1000;
-    this.maximum_requests = maximum_requests;
+  constructor(timeWindow, maximumRequests) {
+    this.timeWindow = timeWindow * 1000;
+    this.maximumRequests = maximumRequests;
     this.userRequestMap = new Map();
   }
 
-  useRateLimiter(userId) {
+  async saveNotificationInteraction(userId, category, notificationId) {
+    await prisma.notificationInteractions.create({
+      data: {
+        userId: userId,
+        timestamp: new Date(),
+        category: category,
+        notificationId: notificationId,
+      },
+    });
+  }
+
+  async getNotificationInteractionsData(userId) {
+    const interactions = await prisma.notificationInteractions.findMany({
+      where: {
+        userId: userId,
+      },
+      orderBy: {
+        timestamp: "desc",
+      },
+    });
+    return interactions;
+  }
+
+  async getUserMostActiveTimeWIndows(userId) {
+    const interactions = await this.getNotificationInteractionsData(userId);
+    const hourCounts = new Array(24).fill(0);
+    interactions.forEach((item) => {
+      const hour = new Date(item.timestamp).getHours();
+      hourCounts[hour]++;
+    });
+
+    const activeHours = hourCounts.map((numberOfClicks, hourIndex) => ({
+      hourIndex,
+      numberOfClicks,
+    }));
+    return activeHours
+      .sort((a, b) => b.numberOfClicks - a.numberOfClicks)
+      .slice(0, 3); //Returns top user's top 3 active hours
```

Comment:
I would use a similar approach that you use used for `getUserPreferredNotificationCategory`, initializing an empty map, looping through the interactions to update the map, sorting, and then taking an array of only the top hour indices (since you don't need the number of clicks).

Reply:
Thanks @azhiah  fixed in https://github.com/izuchukwumba/peered/pull/20/commits/25122476dc6f2cf044ae1f51c3e04fba213dca0d